### ISSUE TITLE
PHPUnit warning: Test code or tested code did not remove its own exception handlers

### ISSUE THREAD

**Issue by nunomaduro** [laravel/framework#49502]
### Laravel Version

Any that uses PHPUnit ^10

### PHP Version

Any

### Description

Seems that, since PHPUnit 10.5.4, running any kind of test suite that boots the Laravel Framework displays the following warning multiple times: `Test code or tested code did not remove its own exception handlers`.

### Steps To Reproduce

```bash
laravel new laravel
cd laravel
./vendor/bin/phpunit
```

**Comment by nunomaduro** [comment]
PHPUnit change that introduced this BC issue: https://github.com/sebastianbergmann/phpunit/pull/5619.

**Comment by rikvdh** [comment]
@nunomaduro Shouldnt this be reported upstream as a regression?

**Comment by nunomaduro** [comment]
@rikvdh It's done. Sebastian is considering reverting it.

**Comment by nunomaduro** [comment]
Reverted. Problem solved.

**Comment by arderyp** [comment]
I never had this issue in 10.x, but I just upgraded from latest 10.x to latest 11.x and 11.0.4 and I'm seeing this on every single test in my project

**Comment by TomBerkGH** [comment]
> I never had this issue in 10.x, but I just upgraded from latest 10.x to latest 11.x and 11.0.4 and I'm seeing this on every single test in my project

i get the same error since laravel update. PHPUnit 11.0.9 and Laravel 11.2.0

**Comment by elb98rm** [comment]
I can confirm this has reappeared:

* L11 : 11.2.0
* PHPUnit : 11.1.1

Looks like it's been re-introduced.

**Comment by oceanapplications** [comment]
> I can confirm this has reappeared:
> 
> * L11 : 11.2.0
> * PHPUnit : 11.1.1
> 
> Looks like it's been re-introduced.

Yup, Sebastian said it would be reintroduced in 11. What's the current work around other than using older phpunit?

**Comment by arderyp** [comment]
various solutions are presented in the Symfony thread: https://github.com/symfony/symfony/issues/53812#issuecomment-1962740145

**Comment by phily245** [comment]
Putting the below in what would be my `Tests/TestCase` worked in the package I am working on and encountered this in:

```php
public function tearDown(): void
{
    parent::tearDown();

    restore_error_handler();
    restore_exception_handler();
}
```

**Comment by zoispag** [comment]
We updated to L11 last week and today Dependabot bumped PHPUnit to v11 and these errors starting to appear for all tests:

- Test code or tested code removed error handlers other than its own
- Test code or tested code removed exception handlers other than its own

As of now, we can't find any working solution. This issue points to Symfony thread which is still open and redirects back to #49622. Does anyone have a working solution? @phily245 suggestion didn't do the trick :(

**Comment by JustChapman** [comment]
We have been digging into this issue at work with a new project. My boss found that removing newRelic in the docker container solved resolved the "Test code or tested code removed..." errors. He also mentioned there appears to be a memory leak in PHPUnit 11.3, though that has not yet been confirmed. I can confirm on my instance, removing newRelic from the docker container did resolve issues for me. Tests execute as expected now.

**Comment by lsfiege** [comment]
> We have been digging into this issue at work with a new project. My boss found that removing newRelic in the docker container solved resolved the "Test code or tested code removed..." errors. He also mentioned there appears to be a memory leak in PHPUnit 11.3, though that has not yet been confirmed. I can confirm on my instance, removing newRelic from the docker container did resolve issues for me. Tests execute as expected now.

I also confirm, removing new relic from my docker container fixed this. But I can't confirm the phpunit memory leak

**Comment by eddyger** [comment]
Reading the phpunit documentation, we can find the #[WithoutErrorHandler] attribute mention on your test method

**Comment by Tofandel** [comment]
Using the suggested

```php

    public function tearDown(): void
    {
        parent::tearDown();

        restore_error_handler();
        restore_exception_handler();
    }
```

I now get the opposite error

`Test code or tested code removed error handlers other than its own`

**Comment by negoziator** [comment]
For those still struggling with this. Adding `HandleExceptions::flushState();` seems to do the trick.

In `tests/TestsCase.php`

Override the `createApplication()` method and add `HandleExceptions::flushState();` before returning `$app`

```php
    /**
     * Creates the application.
     *
     * @return \Illuminate\Foundation\Application
     */
    public function createApplication()
    {
        $app = require Application::inferBasePath().'/bootstrap/app.php';

        $app->make(Kernel::class)->bootstrap();

        HandleExceptions::flushState(); // <--- Add this line

        return $app;
    }
```

**Comment by andreshg112** [comment]
@negoziato, now I have the same warning as @Tofandel:
Test code or tested code removed error handlers other than its own
Test code or tested code removed exception handlers other than its own

**Comment by andreshg112** [comment]
Never mind! Looks like the warning above was caused by this:

![Image](https://github.com/user-attachments/assets/d1c56011-56c6-4168-a01a-75b9d111b915)

**Comment by andreshg112** [comment]
Not really. I still see a lot of those warnings even after removing the code above.

**Comment by specialtactics** [comment]
It seems like this is back in 11.

**Comment by cdburgess** [comment]
I'm getting the same issue. Here is some more information in case it helps someone:

**phpunit.xml**
```
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         bootstrap="tests/bootstrap.php"
         colors="true">
```

**bootstrap.php**
```
// Create application
$app = require __DIR__ . '/../bootstrap/app.php';
$app->make(Illuminate\Contracts\Console\Kernel::class)->bootstrap();
HandleExceptions::flushState();
```

The HandleExceptions::flushState() (taken from a comment above) seems to do the trick.

**ISSUE VERSION**
PHPUnit 12.0.5

**Comment by marcobax** [comment]
I'm using 

```
"laravel/framework": "^12.16.0",
"phpunit/phpunit": "^11.5.15",
"nunomaduro/collision": "^8.8.0",
```
and keep getting the "Test code or tested code removed error handlers other than its own" error for each test.
I tried adding this code to `createApplication()`
`HandleExceptions
::flushState();`
but it doesn't work unfortunately.

**Comment by WalkWeb** [comment]
Following the advice above, I added

```php
    public function tearDown(): void
    {
        parent::tearDown();

        restore_error_handler();
        restore_exception_handler();
    }
```

but I started getting an error:

`Test code or tested code removed error handlers other than its own`

This code helped me:

```php
    public function tearDown(): void
    {
        parent::tearDown();

        restore_exception_handler();
    }
```

**Comment by amcsi** [comment]
I'm confused. What am I supposed to do?

I tried the `HandleExceptions::flushState();` bit and the [tearDown](https://github.com/laravel/framework/issues/49502#issuecomment-2948464945) bit as well both together and separately as well, and neither does anything.

There's nothing in the Laravel docs about what to do here, nor is there anything in the PhpUnit docs. I'm stuck and am not able to upgrade to Laravel 12 this way.

**Comment by amcsi** [comment]
I've figured it out:

The issue was stemming by the fact that I had a custom PhpUnit bootstrap class I was using, and in it I was doing a `createApplication()` (i.e. `$app->make(Kernel::class)->bootstrap()`) to be able to call some artisan commands before/after running the tests.

I simply had to make sure I ran `HandleExceptions::flushState();` at the end of each method where I was doing `createApplication()` in that boostrap class.

Probably what was happening was that if I manually create an application outside of tests, it would add extra error/exception handlers that neither Laravel nor PhpUnit would know about, and then PhpUnit would complain about it in the end.

**Comment by marcobax** [comment]
> I've figured it out:
> 
> The issue was stemming by the fact that I had a custom PhpUnit bootstrap class I was using, and in it I was doing a `createApplication()` (i.e. `$app->make(Kernel::class)->bootstrap()`) to be able to call some artisan commands before/after running the tests.
> 
> I simply had to make sure I ran `HandleExceptions::flushState();` at the end of each method where I was doing `createApplication()` in that boostrap class.
> 
> Probably what was happening was that if I manually create an application outside of tests, it would add extra error/exception handlers that neither Laravel nor PhpUnit would know about, and then PhpUnit would complain about it in the end.

This works for me, massive thanks!

**Comment by LukeTowers** [comment]
Not sure if anyone else will find this helpful, but I discovered the following:

### TL;DR:
If a test is marked as skipped and your `tearDown()` throws an exception before Laravel’s cleanup runs, PHPUnit will swallow that teardown exception and instead show the vague “Test code or tested code did not remove its own exception handlers” message.

### Why:
`markTestSkipped()` throws a `SkippedWithMessageException`.
PHPUnit’s `runBare()` always calls `tearDown()` afterward, but if `tearDown()` throws, the exception is ignored because PHPUnit assumes the earlier skip exception is the “real” outcome.
Since Laravel’s cleanup never ran, global handlers remain, triggering the risky-test warning.

### Example:

```php
public function setUp(): void
{
    parent::setUp();

    if (!File::exists(base_path('node_modules'))) {
        $this->markTestSkipped('This test requires node_modules to be installed');
    }

    // Never reached if skipped:
    $this->fixturePath = base_path('modules/system/tests');
}

public function tearDown(): void
{
    // Uses $this->fixturePath but it’s null if skipped early
    if (File::isDirectory($this->fixturePath . '/node_modules')) {
        File::deleteDirectory($this->fixturePath . '/node_modules');
    }

    parent::tearDown();
}
```

### Result:
Instead of the teardown exception, PHPUnit only reports:

> Test code or tested code did not remove its own exception handlers.

See https://github.com/sebastianbergmann/phpunit/pull/6281

**Comment by crmpicco** [comment]
@WalkWeb 's second workaround seems to work for me. I'm on the latest from the 11 branch.

```
bin/phpunit --version
PHPUnit 11.5.31 by Sebastian Bergmann and contributors.
```

**Comment by php-bob** [comment]
For me, once I added the  `parent::tearDown()` the error went away.  Since I didn't really need my 'custom' `tearDown()` I just removed it.
And all is well.

**Comment by pallad42** [comment]
PHPUnit 11.5
PHP 8.2
Symfony 6.4

The suggested solutions didn’t work for me, but I found a way to bypass the problem. The only solution that worked was to mark each method in `KernelTestCase` and `WebTestCase` with the `#[RunInSeparateProcess]` attribute. This way, each test runs in a separate, isolated process, so the handlers from previous tests are not overwritten. I’m not sure if this solution is correct or efficient, but I spent many hours debugging and couldn’t come up with anything better.

So instead of using `dama/doctrine-test-bundle`, I handle transaction rollbacks manually. I think DAMA doesn’t work properly when tests are executed in isolated processes.

KernelTestCase
```php
protected function setUp(): void
{
    parent::setUp();
    $kernel = self::bootKernel();

    $this->entityManager = $kernel->getContainer()->get('doctrine')->getManager();

    $this->entityManager->beginTransaction();
    $this->entityManager->getConnection()->setAutoCommit(false);
}

protected function tearDown(): void
{
    parent::tearDown();

    if ($this->entityManager->getConnection()->isTransactionActive()) {
        $this->entityManager->rollback();
    }

    $this->entityManager->close();
}

#[RunInSeparateProcess]
public function testSomething() { ... } // test method will run inside a DB transaction and roll back at the end
```

WebTestCase
Unfortunately, neither `DAMA` nor the `KernelTestCase` approach worked in this case. The only solution I found was to use the following (it is not efficient way, but it works):
```php
protected function setUp(): void
{
    parent::setUp();
    $this->client = static::createClient();
    $this->entityManager = static::getContainer()->get(EntityManagerInterface::class);
    exec('php bin/console doctrine:fixtures:load --no-interaction --env=test');
}

#[RunInSeparateProcess]
public function testSomething() { ... } // database will be overwritten with fixtures before each test
```
Maybe consider experimenting with the `liip/test-fixtures-bundle` package — it might work, although I haven’t tested it.

**Comment by lukemorcom** [comment]
> I've figured it out:
> 
> The issue was stemming by the fact that I had a custom PhpUnit bootstrap class I was using, and in it I was doing a `createApplication()` (i.e. `$app->make(Kernel::class)->bootstrap()`) to be able to call some artisan commands before/after running the tests.
> 
> I simply had to make sure I ran `HandleExceptions::flushState();` at the end of each method where I was doing `createApplication()` in that boostrap class.
> 
> Probably what was happening was that if I manually create an application outside of tests, it would add extra error/exception handlers that neither Laravel nor PhpUnit would know about, and then PhpUnit would complain about it in the end.

Same for me too despite the `createApplication()` call being in an entirely different test class to where I was seeing the error.

Many thanks!
